### PR TITLE
Performance 4 (plugins)

### PR DESCRIPTION
**Pull request by frankmayer**
**[UPDATED (1.6.2017) - see third paragraph]**
Performance (and a bit of cleanup) Batch #4

The changes in this batch are all in files under `plugins`.
This PR modifies code to be a bit more performant and also does some cleanup.

I have mostly done work on low hanging fruit. There are still other ways of improving, but whose involve more deep research in the code and probably more drastic changes in order to be implemented.

**[Update 6.1.2017]**
In order to lighten this PR up, I have introduced 5 sub-PRs, which are easier to digest and mostly with very specific changes. In order to continue with this one, those ones should be tested/reviewed/merged first. After that, I can resolve the conflicts and there should finally be only a few changes left.

- [x] 1) Replaced unnecessary double quotes in plugins #13211
- [x] 2) Removed unnecessary parentheses in plugins #13212
- [x] 3) Various changes in plugins #13213
- [x] 4) Type safe string comparison of strings in plugins #13271
- [x] 5) Type safe comparison in plugins - second iteration #13272

**[Update 1.6.2017]**
All Sub-PRs of this PR have now been merged. That means a lot less changes to be reviewed :smile:
Please review. 
Thank you!

### CODE REVIEW

File: plugins/captcha/recaptcha/recaptchalib.php [hunk 1]
```diff
@@ -53,10 +53,10 @@ class JReCaptcha
 	 */
 	public function __construct($secret)
 	{
-		if ($secret == null || $secret == "")
+		if ($secret === null || $secret === '')
 		{
 			die("To use reCAPTCHA you must get an API key from <a href='"
-				. self::$_signupUrl . "'>" . self::$_signupUrl . "</a>");
+				. self::$_signupUrl . "'>" . self::$_signupUrl . '</a>');
```
Comment (andrepereiradasilva) [inline plugins/captcha/recaptcha/recaptchalib.php]: again why use `'` when you use `"` in all others?

File: plugins/content/pagenavigation/pagenavigation.php [hunk 2]
```diff
@@ -39,7 +39,7 @@ public function onContentBeforeDisplay($context, &$row, &$params, $page = 0)
 			return false;
 		}
 
-		if (($context == 'com_content.article') && ($view == 'article') && $params->get('show_item_navigation'))
+		if (($context === 'com_content.article') && ($view === 'article') && $params->get('show_item_navigation'))
```
Comment (andrepereiradasilva) [inline plugins/content/pagenavigation/pagenavigation.php]: you could remove the parenthisis here

File: plugins/user/profile/profile.php [hunk 3]
```diff
@@ -374,15 +374,12 @@ public function onUserBeforeSave($user, $isnew, $data)
 		$task       = JFactory::getApplication()->input->getCmd('task');
 		$option     = JFactory::getApplication()->input->getCmd('option');
 		$tosarticle = $this->params->get('register_tos_article');
-		$tosenabled = ($this->params->get('register-require_tos', 0) == 2) ? true : false;
+		$tosenabled = ($this->params->get('register-require_tos', 0) == 2);
 
-		if (($task == 'register') && ($tosenabled) && ($tosarticle) && ($option == 'com_users'))
+		// Check that the tos is checked.
+		if (($task === 'register') && $tosenabled && $tosarticle && ($option === 'com_users') && !$data['profile']['tos'])
```
Comment (andrepereiradasilva) [inline plugins/user/profile/profile.php]: you could remove the parenthisis here

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on 42a5a54b003afa2e4bc86af7f9035fc8de52dcff<br /><br />code review<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12228">issues.joomla.org/tracker/joomla-cms/12228</a>.</sub>

File: plugins/system/debug/debug.php [hunk 4]
```diff
@@ -919,12 +915,12 @@ protected function displayQueries()
 			// Increment the count.
 			if (stripos($query, 'select') === 0)
 			{
-				$selectQueryTypeTicker[$fromString] = $selectQueryTypeTicker[$fromString] + 1;
+				++$selectQueryTypeTicker[$fromString];
```
Comment (csthomas) [inline plugins/system/debug/debug.php]: I would better use `$selectQueryTypeTicker[$fromString] += 1;` which is more readable.

File: plugins/system/redirect/redirect.php [hunk 5]
```diff
@@ -233,7 +233,7 @@ private static function doErrorHandling($error)
 		// We have an unpublished redirect object, increment the hit counter
 		else
 		{
-			$redirect->hits += 1;
+			++$redirect->hits;
```
Comment (csthomas) [inline plugins/system/redirect/redirect.php]: Please revert it. This is only my personal advice.

File: plugins/finder/contacts/contacts.php [hunk 6]
```diff
@@ -251,7 +251,7 @@ public function onFinderChangeState($context, $pks, $value)
 	protected function index(FinderIndexerResult $item, $format = 'html')
 	{
 		// Check if the extension is enabled
-		if (JComponentHelper::isEnabled($this->extension) == false)
+		if (JComponentHelper::isEnabled($this->extension) === false)
```
Comment (csthomas) [inline plugins/finder/contacts/contacts.php]: the same

File: plugins/finder/content/content.php [hunk 7]
```diff
@@ -246,7 +246,7 @@ protected function index(FinderIndexerResult $item, $format = 'html')
 		$item->setLanguage();
 
 		// Check if the extension is enabled.
-		if (JComponentHelper::isEnabled($this->extension) == false)
+		if (JComponentHelper::isEnabled($this->extension) === false)
```
Comment (csthomas) [inline plugins/finder/content/content.php]: the same

File: plugins/finder/newsfeeds/newsfeeds.php [hunk 8]
```diff
@@ -252,7 +252,7 @@ public function onFinderChangeState($context, $pks, $value)
 	protected function index(FinderIndexerResult $item, $format = 'html')
 	{
 		// Check if the extension is enabled.
-		if (JComponentHelper::isEnabled($this->extension) == false)
+		if (JComponentHelper::isEnabled($this->extension) === false)
```
Comment (csthomas) [inline plugins/finder/newsfeeds/newsfeeds.php]: the same

File: plugins/finder/tags/tags.php [hunk 9]
```diff
@@ -206,7 +206,7 @@ public function onFinderChangeState($context, $pks, $value)
 	protected function index(FinderIndexerResult $item, $format = 'html')
 	{
 		// Check if the extension is enabled
-		if (JComponentHelper::isEnabled($this->extension) == false)
+		if (JComponentHelper::isEnabled($this->extension) === false)
```
Comment (csthomas) [inline plugins/finder/tags/tags.php]: the same

File: plugins/content/contact/contact.php [hunk 10]
```diff
@@ -98,7 +98,7 @@ protected function getContactId($created_by)
 		$query->where('contact.published = 1');
 		$query->where('contact.user_id = ' . (int) $created_by);
 
-		if (JLanguageMultilang::isEnabled() == 1)
+		if (JLanguageMultilang::isEnabled() === 1)
```
Comment (frankmayer) [inline plugins/content/contact/contact.php]: Thanks a lot, this one slipped through... :) 
I will go for `if (JLanguageMultilang::isEnabled())`.

File: plugins/finder/categories/categories.php [hunk 11]
```diff
@@ -244,7 +244,7 @@ public function onFinderChangeState($context, $pks, $value)
 	protected function index(FinderIndexerResult $item, $format = 'html')
 	{
 		// Check if the extension is enabled.
-		if (JComponentHelper::isEnabled($this->extension) == false)
+		if (JComponentHelper::isEnabled($this->extension) === false)
```
Comment (csthomas) [inline plugins/finder/categories/categories.php]: It is OK

File: plugins/system/debug/debug.php (hunk 4 above)
Comment (frankmayer) [inline plugins/system/debug/debug.php]: Hm, yes, since it is not directly assigned to anything, we can do that. Good idea. Will do.

File: plugins/finder/categories/categories.php (hunk 11 above)
Comment (csthomas) [inline plugins/finder/categories/categories.php]: Why have you not changed that line?
Remember to resolve conflicts in files.

File: plugins/authentication/ldap/ldap.php [hunk 12]
```diff
@@ -70,7 +70,7 @@ public function onUserAuthenticate($credentials, $options, &$response)
 			{
 				// Bind using Connect Username/password
 				// Force anon bind to mitigate misconfiguration like [#7119]
-				if (strlen($this->params->get('username')))
+				if ($this->params->get('username') !== '')
```
Comment (csthomas) [inline plugins/authentication/ldap/ldap.php]: `$this->params` is an intance of Registry. Method `get` if key not exists may returns `null`.
I would only change to `if ($this->params->get('username'))`.
Otherwise you have to be 100% sure that this won't break anything.

File: plugins/captcha/recaptcha/recaptchalib.php [hunk 13]
```diff
@@ -132,7 +131,7 @@ public function verifyResponse($remoteIp, $response)
 		$answers = json_decode($getResponse, true);
 		$recaptchaResponse = new JReCaptchaResponse();
 
-		if (trim($answers['success']) == true)
+		if (trim($answers['success']) === true)
```
Comment (csthomas) [inline plugins/captcha/recaptcha/recaptchalib.php]: I hope this is my last error note :)

If am I right function `trim` returns string.
I would replace by `if (trim($answers['success']))`

After you fix it I can mark test as success.

Comment (frankmayer) [inline plugins/captcha/recaptcha/recaptchalib.php]: Yes, I was being overzealous... :smile:  Will do.

**Comment by csthomas** [comment]
I have tested this item :white_check_mark: successfully on 475b88019db96f67bec8a075edfe357a72561db7<br /><br />Code review.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12228">issues.joomla.org/tracker/joomla-cms/12228</a>.</sub>

**Comment by wilsonge** [comment]
All the sub PR's are now merged. Do you want to resolve conflicts here for the last few things?

**Comment by frankmayer** [comment]
@wilsonge thanks, yes, will do

**Comment by frankmayer** [comment]
Conflicts resolved...

**Comment by frankmayer** [comment]
Hmm, this has still lots of different changes. Will do one or more sub PR's to lighten this up a little more.

File: plugins/system/fields/fields.php [hunk 14]
```diff
@@ -472,6 +472,8 @@ public function onContentPrepare($context, $item)
 		{
 			$item->fields[$field->id] = $field;
 		}
+
+		return;
```
Comment (wilsonge) [inline plugins/system/fields/fields.php]: This isn't needed?

**Comment by frankmayer** [comment]
Edited initial post to show which sub-PRs still have to go through, in order to make this one easier to digest.
It would be great if the initial big batch PRs of mine could make it into alpha2. Thanks!

**Comment by frankmayer** [comment]
Pls don't review this yet. There was a git problem. Need to resolve first.

**Comment by frankmayer** [comment]
All Sub-PRs of this PR have now been merged. That means a lot less changes to be reviewed :smile:
Please review. 
Thank you!

**Comment by andrepereiradasilva** [comment]
next in my list of tests - tomorrow probably

File: plugins/captcha/recaptcha/recaptcha.php [hunk 15]
```diff
@@ -268,7 +268,7 @@ private function _recaptcha_http_post($host, $path, $data, $port = 80)
 
 		$response = '';
 
-		if (($fs = @fsockopen($host, $port, $errno, $errstr, 10)) == false )
+		if (($fs = @fsockopen($host, $port, $errno, $errstr, 10)) === false )
```
Comment (Quy) [inline plugins/captcha/recaptcha/recaptcha.php]: Remove space after `false`.

File: plugins/content/pagenavigation/pagenavigation.php [hunk 16]
```diff
@@ -39,7 +39,7 @@ public function onContentBeforeDisplay($context, &$row, &$params, $page = 0)
 			return false;
 		}
 
-		if (($context === 'com_content.article') && ($view === 'article') && $params->get('show_item_navigation'))
+		if (($context === 'com_content.article') && $view === 'article' && $params->get('show_item_navigation'))
```
Comment (Quy) [inline plugins/content/pagenavigation/pagenavigation.php]: Remove parenthesis `($context === 'com_content.article')`

File: plugins/system/debug/debug.php [hunk 17]
```diff
@@ -1110,7 +1106,7 @@ protected function displayQueries()
 					$title = '<span class="dbg-noprofile">' . $title . '</span>';
 				}
 
-				$htmlProfile = ($info[$id]->profile ? $info[$id]->profile : JText::_('PLG_DEBUG_NO_PROFILE'));
+				$htmlProfile = ($info[$id]->profile ?: JText::_('PLG_DEBUG_NO_PROFILE'));
```
Comment (Quy) [inline plugins/system/debug/debug.php]: Remove outside parenthesis (ignore if done in a different PR).

Comment (frankmayer) [inline plugins/system/debug/debug.php]: Don't remember if that is in any other PR. There are so many PR's still open, that some may overlap. I'll just do that here and move on :smile: 
Thanks for reviewing!

File: plugins/user/profile/profile.php [hunk 18]
```diff
@@ -384,13 +384,10 @@ public function onUserBeforeSave($user, $isnew, $data)
 		$tosarticle = $this->params->get('register_tos_article');
 		$tosenabled = ($this->params->get('register-require_tos', 0) == 2);
 
-		if (($task === 'register') && $tosenabled && $tosarticle && ($option === 'com_users'))
+		// Check that the tos is checked.
+		if (($task === 'register') && $tosenabled && $tosarticle && $option === 'com_users' && !$data['profile']['tos'])
```
Comment (Quy) [inline plugins/user/profile/profile.php]: Remove parenthesis `($task === 'register')`.

**Comment by frankmayer** [comment]
@Quy would you be so kind and check the changes of your reviews for this one, too? So maybe we can get this RTC as soon as @andrepereiradasilva also has time to check it?
Thanks guys!!!

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 475b88019db96f67bec8a075edfe357a72561db7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12228">issues.joomla.org/tracker/joomla-cms/12228</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
RTC after two successful tests.

**Review by andrepereiradasilva — COMMENTED** [review]
some minor issues/sugestions

**Comment by franz-wohlkoenig** [comment]
@andrepereiradasilva set Status back on Pending?

**Comment by frankmayer** [comment]
@franz-wohlkoenig Let me just fly over those comments and I'll come back to you.

File: plugins/authentication/ldap/ldap.php [hunk 19]
```diff
@@ -72,7 +72,7 @@ public function onUserAuthenticate($credentials, $options, &$response)
 			{
 				// Bind using Connect Username/password
 				// Force anon bind to mitigate misconfiguration like [#7119]
-				if (strlen($this->params->get('username')))
+				if ($this->params->get('username'))
```
Comment (frankmayer) [inline plugins/authentication/ldap/ldap.php]: yes :+1:

Comment (frankmayer) [inline plugins/authentication/ldap/ldap.php]: Yes.. thanks

File: plugins/captcha/recaptcha/recaptcha.php [hunk 20]
```diff
@@ -40,7 +40,7 @@ public function onInit($id = 'dynamic_recaptcha_1')
 	{
 		$pubkey = $this->params->get('public_key', '');
 
-		if ($pubkey == null || $pubkey == '')
+		if ($pubkey === null || $pubkey === '')
```
Comment (frankmayer) [inline plugins/captcha/recaptcha/recaptcha.php]: I think it can't... changing

**Comment by frankmayer** [comment]
@franz-wohlkoenig @andrepereiradasilva updated according to reviewer's comments. 
Please check and RTC.
Thanks!!

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on 475b88019db96f67bec8a075edfe357a72561db7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12228">issues.joomla.org/tracker/joomla-cms/12228</a>.</sub>
